### DISCUSSION TITLE
Can not get SNR values

### DISCUSSION THREAD

**Discussion by vemanthchandrap** [sandeepmistry/arduino-LoRa#671] [Q&A]
Hello everyone,
I am doing my first project on LoRa and I am using two Dragino LoRa shields (868Mhz) with sx1276 and two Arduino uno wifi rev2 as transmitter and receiver. I am using Sandeep Mistry LoRa library to get RSSI and SNR values. I am able to get RSSI values but I am not getting values of SNR values. any suggestions why I am not able to get SNR values.

code below
for receiver
```
#include <SPI.h>
#include <LoRa.h>

#define BAND 868E6 
#define SS_PIN 10
#define RST_PIN 9
#define DI0 2

void setup() {
  Serial.begin(9600);
  while (!Serial);

  if (!LoRa.begin(BAND)) {
    Serial.println("LoRa init failed. Check your connections.");
    while (1);
  }

  LoRa.onReceive(onReceive);
  LoRa.receive();
}

void loop() {

}

void onReceive(int packetSize) {
  if (packetSize == 0) return;

  while (LoRa.available()) {
    String data = LoRa.readString();
    Serial.print("Received: ");
    Serial.println(data);
  }

  int rssi = LoRa.packetRssi();
  int snr = LoRa.packetSnr();
  Serial.print("RSSI: ");
  Serial.print(rssi);
  Serial.print(" dBm, SNR: ");
  Serial.print(snr);
  Serial.println(" dB");
}
```

**Comment by morganrallen** [comment]
What value does `packetSnr()` return?

**Reply by morganrallen** [reply]
I think there is a bug in the `packetSnr` method. I'm reading the datasheet at here is the registers description.

> Estimation of SNR on last packet received. In __two’s compliment__ format mutiplied by 4.

And here's the source of the method...
```
return ((int8_t)readRegister(REG_PKT_SNR_VALUE)) * 0.25;
```

It should be converting the value from the register to a negative number with two's complement but it's not.

**Reply by vemanthchandrap** [reply]
I really appreciate your help. I think you are right I will check this tomorrow. Thank you so much.

**Reply by vemanthchandrap** [reply]
Hey Morgan, now I understood the problem. Instead of this  < return ((int8_t)readRegister(REG_PKT_SNR_VALUE)) * 0.25; >  i need a code that converts the value from the register number with two's complement. Could you please send me the correct code instead of the error? That would be really appreciated. thank you in advance.

**Reply by morganrallen** [reply]
I'm working on getting a patch together but it might result in some non-trivial changes

**Reply by vemanthchandrap** [reply]
Okay no problem, Let me know when the actual work is done. I will be looking forward to it.

**Reply by morganrallen** [reply]
also you didn't answer my initial question, what value did you get?

**Reply by vemanthchandrap** [reply]
Nothing, I am not getting anything from packetSnr() in this library so the values of SNR are not shown,  but I am getting RSSI values. I would like to share my transmitter and receiver codes.
TRANSMITTER CODE
```
#include <SPI.h>
#include <LoRa.h>

  void setup() {
  Serial.begin(9600);
  while (!Serial);

  if (!LoRa.begin(868E6)) {
    Serial.println("LoRa initialization failed. Check your connections.");
    while (1);
  }
}

  void loop() {
  String message = "Hello!";

  LoRa.beginPacket();
  LoRa.print(message);
  LoRa.endPacket();

  Serial.println("Message sent: " + message);
  delay(5000); 
}
```
RECEIVER CODE

```
#include <SPI.h>
#include <LoRa.h>

void setup() {
  Serial.begin(9600);
  while (!Serial);

 if (!LoRa.begin(868E6)) {
    Serial.println("LoRa initialization failed. Check your connections.");
    while (1);
  }
}
   void loop() {
   int packetSize = LoRa.parsePacket();
   if (packetSize) {
   String receivedData = "";
   while (LoRa.available()) {
    receivedData += (char)LoRa.read();
    }

    Serial.println("Received: " + receivedData);
     int rssi = LoRa.packetRssi();
    float snr = LoRa.packetSnr();
     
    Serial.print("RSSI: ");
    Serial.println(rssi);
    Serial.print("SNR: ");
    Serial.println(snr);
    }
    }
```
I am using these codes and trying to get the data.

**Reply by morganrallen** [reply]
I've pushed a new branch that handles the two's complement of the SNR register. This also changes the return value from `float` to `int` which should solve some other issues I saw while looking into this. There's also some naming changes that might want to be brought in too. 

Try checking out this branch and giving it another go. I get `+9dB` with two devices on my desk, seems like a reasonable result.
https://github.com/sandeepmistry/arduino-LoRa/tree/twos_snr

Specifically...
The datasheet refers to the register as `packetSnr` but refers to the results of `packetSnr / 4` as `SNR[dB]` so technically a method called `packetSnr` should only return the raw result, and maybe `SNR` return the divided result. This could address #570 while we're at it. Thoughts @sandeepmistry ?

**Reply by vemanthchandrap** [reply]
Thank you, I have tried changing the branch in LoRa.cpp and LoRa.h and I changed the code from float to int for SNR but there I can't find SNR values in the serial monitor. I think there is some problem on my side. I have uploaded a few pictures you can look at them.
![snr1](https://github.com/sandeepmistry/arduino-LoRa/assets/148960361/d3338e8b-4841-4e40-823a-925f3ffe876a)
![snr2](https://github.com/sandeepmistry/arduino-LoRa/assets/148960361/df59b21b-5487-4664-a0a1-266757573ae6)
![snr3](https://github.com/sandeepmistry/arduino-LoRa/assets/148960361/5ed3f2ab-0207-4eb1-b69b-b73e1c2543e3)
![snr4](https://github.com/sandeepmistry/arduino-LoRa/assets/148960361/903caeff-5874-4a54-ae5f-adc3fa235fc5)

**Reply by morganrallen** [reply]
Hmm, the output in the log doesn't seem to match what I would expect from that code.

`Received: ` vs `Received Packet '...'`
`RSSI: ...` vs `RSSI -13`

Check your code again and make sure that's what's running. Additionally when printing the SNR try `Serial.printf("SNR: %d\n", snr);`

**Reply by vemanthchandrap** [reply]
I suppose that Serial.printf is not included in LoRa.h because I am getting an error.

**Reply by morganrallen** [reply]
If you don't post the error no one can help you.

**Reply by vemanthchandrap** [reply]
I have only included SPI.h and LoRa.h libraries. and I have got this error for  serialprintf() 

> C:\Users\Vemanth chandra\AppData\Local\Temp\.arduinoIDE-unsaved2023927-16316-yartv2.hwkhe\sketch_oct27c\sketch_oct27c.ino: In function 'void loop()':
C:\Users\Vemanth chandra\AppData\Local\Temp\.arduinoIDE-unsaved2023927-16316-yartv2.hwkhe\sketch_oct27c\sketch_oct27c.ino:31:12: error: 'class UartClass' has no member named 'printf'; did you mean 'print'?
     Serial.printf("SNR: %d\n", snr);
            ^~~~~~
            print
exit status 1
Compilation error: 'class UartClass' has no member named 'printf'; did you mean 'print'?

**Comment by morganrallen** [comment]
Odd, I thought that was a standard part of the Serial class.....
Regardless, you need to double check your code as the log output does not
seem to match.

**Reply by vemanthchandrap** [reply]
I have checked the code I think RSSI values seems fine for me because if I am changing the position of transmitter and receiver the values are increasing and decreasing I don't see any problem in that, but I think the problem is with SNR I don't see in that.

**Reply by vemanthchandrap** [reply]
@sandeepmistry any thoughts on my issue?

**Reply by vemanthchandrap** [reply]
@morganrallen I just want to know how you got +9dB for SNR values whereas I am not even getting SNR values.

**Reply by morganrallen** [reply]
As I said before, the code snippet you posted and the log output do not seem to match, I suspect you're not running the code you think you are. Do you have multiple devices connected? Make sure the one you are flashing is the same as the one you're monitoring.

**Reply by vemanthchandrap** [reply]
I am working with two Lora devices connected to my laptop. I have checked it, I am monitoring the right one (receiver), and if you could check and you feel anything is wrong please correct the codes I posted yesterday it would be helpful. Thank you in advance.

**Reply by morganrallen** [reply]
So not really sure what to tell you, I tried your code but had to modify it to not call `Serial.print` inside `onReceive` and it worked as expected. Here's my variation for a specific ESP32 based board.

```
#include <SPI.h>
#include <LoRa.h>

void setup() {
  Serial.begin(115200);
  while (!Serial);

  Serial.println("LoRa Receiver");

  LoRa.setPins(15, 4, 5);

  if (!LoRa.begin(915E6)) {
    Serial.println("Starting LoRa failed!");
    while (1);
  }

  LoRa.onReceive(onReceive);
  LoRa.receive();
}

bool data = false;

void loop() {
  if (data) {
    while (LoRa.available()) {
      String data = LoRa.readString();
      Serial.print("Received: ");
      Serial.println(data);
    }

    int rssi = LoRa.packetRssi();
    int snr = LoRa.packetSnr();
    Serial.print("RSSI: ");
    Serial.print(rssi);
    Serial.print(" dBm, SNR: ");
    Serial.print(snr);
    Serial.println(" dB");
    data = false;
  }
}

void onReceive(int packetSize) {
  // handle data inside loop to avoid ISR issues
  data = true;
}
```

and here's my output

```
Received: �r
            [�r���@�����ψ�BU�j2ő��U���w���{�G▒ʙ>߿XXF'�Ņbs�ϫ�9�1:Ma��-��Tp��     uѯA��Є�o4��.!Pd��8؝�NVב,
0���
RSSI: -56 dBm, SNR: 9 dB

```

**Reply by vemanthchandrap** [reply]
It seems like I am unable to get SNR values with my microcontroller(ARDUINO UNO WIFI REV2) maybe I will try with ESP32 or any other broad that can retrieve SNR values. Thank you very much @morganrallen for helping me. It means a lot. I think I have a solution now.

**Reply by vemanthchandrap** [reply]
Hey @morganrallen,  hope you are doing good. this is my first project. i have codes for transmitter and receiver and i am using LoRa library by sandeep mistry in arduino ide. I used TTGO LoRa32 OLED as my board and i didn't get any error in the output, but i can't see any RSSI and SNR output values in serial monitor. anyone can help me with this. Please help me I am clueless, any chances I need to do to get the values on serial monitor. I used the codes from https://randomnerdtutorials.com/ttgo-lora32-sx1276-arduino-ide/

output:
Sketch uses 280941 bytes (21%) of program storage space. Maximum is 1310720 bytes.
Global variables use 22388 bytes (7%) of dynamic memory, leaving 272524 bytes for local variables. Maximum is 294912 bytes.
esptool.py v4.5.1
Serial port COM11
Connecting....
Chip is ESP32-D0WDQ6-V3 (revision v3.1)
Features: WiFi, BT, Dual Core, 240MHz, VRef calibration in efuse, Coding Scheme None
Crystal is 40MHz
MAC: 34:98:7a:6b:f7:a0
Uploading stub...
Running stub...
Stub running...
Changing baud rate to 921600
Changed.
Configuring flash size...
Flash will be erased from 0x00001000 to 0x00005fff...
Flash will be erased from 0x00008000 to 0x00008fff...
Flash will be erased from 0x0000e000 to 0x0000ffff...
Flash will be erased from 0x00010000 to 0x00054fff...
Compressed 17568 bytes to 12204...
Writing at 0x00001000... (100 %)
Wrote 17568 bytes (12204 compressed) at 0x00001000 in 0.5 seconds (effective 295.2 kbit/s)...
Hash of data verified.
Compressed 3072 bytes to 146...
Writing at 0x00008000... (100 %)
Wrote 3072 bytes (146 compressed) at 0x00008000 in 0.1 seconds (effective 391.8 kbit/s)...
Hash of data verified.
Compressed 8192 bytes to 47...
Writing at 0x0000e000... (100 %)
Wrote 8192 bytes (47 compressed) at 0x0000e000 in 0.1 seconds (effective 493.3 kbit/s)...
Hash of data verified.
Compressed 281312 bytes to 152630...
Writing at 0x00010000... (10 %)
Writing at 0x0001d278... (20 %)
Writing at 0x000250b1... (30 %)
Writing at 0x0002a328... (40 %)
Writing at 0x0002f84c... (50 %)
Writing at 0x00034dac... (60 %)
Writing at 0x0003d6a4... (70 %)
Writing at 0x0004812f... (80 %)
Writing at 0x0004d6ff... (90 %)
Writing at 0x00052ebb... (100 %)
Wrote 281312 bytes (152630 compressed) at 0x00010000 in 2.6 seconds (effective 881.1 kbit/s)...
Hash of data verified.

Leaving...
Hard resetting via RTS pin...
